### DISCUSSION TITLE
InvalidAddressError: Address "undefined" is invalid.

### DISCUSSION THREAD

**Discussion by Neaxic** [wevm/wagmi#4284] [Q&A] [answered]
Im hitting this "InvalidAddressError: Address "undefined" is invalid." error when trying to use the writeContract function, from the useWriteContract hook. I have a simmilar hook to approve erc721, that that works just fine. - i for some reason have issues with this hook, that interacting with my own contract - but i have no clue why.

Ive tried troubleshooting this quite a bit, when just doing the writecontract i get no response at all, nothing happens - when i take the error value out from the useWriteContract hook, i get the above defined error.

Im curretly using 
    "viem": "^2.21.9",
    "wagmi": "^2.12.12"
    "@tanstack/react-query": "^5.55.0",
    "@rainbow-me/rainbowkit": "^2.1.5",

Below is a combi of all relevant code - if theres anything missing let me know.

My write hook to my contract
`
 const {
    writeContract,
    error,
    isError,
    isPending,
  } = useWriteContract();
 
 ...code...

writeContract({
      abi: switchContractABI,
      address: '0x8a901fb2dBfE727279C3738a4FF7Da61843d2853',
      functionName: 'donateBatchNFT',
      args: [['0x9982d1e50059bc44e92f7084f532081c12c0bf6a', 4, 0, true]],
    });
`

The sol contract function im trying to invoke:
`
struct DonationInput {
        address nftContract;
        uint256 tokenId;
        uint256 donorFee;
        bool isERC721;
    }
function donateBatchNFT(DonationInput[] calldata _donations) external`

The abi for that function - this is stored in a seperate json file that is imported 
`{
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "nftContract",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "donorFee",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isERC721",
            "type": "bool"
          }
        ],
        "internalType": "struct Switch.DonationInput[]",
        "name": "_donations",
        "type": "tuple[]"
      }
    ],
    "name": "donateBatchNFT",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },`
    
    
Anybody hitting similar issues? I find it very odd my approval hook works, yet this dosent. - When interacting with remix IDE directly, i can invoke the function with an identical arg list.

![image](https://github.com/user-attachments/assets/7fec9416-8277-40c0-b8bc-e0c472cba373)

**Comment by tmm** [comment, marked as answer]
Are you using TypeScript? If not, would highly recommend using it so [`args` is automatically inferred](https://wagmi.sh/react/typescript#const-assert-abis-typed-data).

Based on the code you shared, `args` is incorrect. The following code should work.

[TypeScript Playground](https://www.typescriptlang.org/play/?target=99&ssl=42&ssc=1&pln=1&pc=1#code/JYWwDg9gTgLgBAbzgVwM4FMDqVg3QYQgDsYoBDAY3gF84AzKCEOAcgHcyBzEYFgWABQgisVTwkbHHkIlyVOLQC8KDNlwFipSjAAUASkHDR8VG1wUAFjK1UAggCEAknGUBtQXEQfPcYETDIMKgAXHDuAj4+CN6RniLgxOgkIWExsZ5IfnhQRGQANgAqAJ5g6KEsZAAmlVDoqKgsADRwuSBlrER0MNZyME1wMCXtFdW19SwKjWnpmSToOfnFpeXIWQBMAKwAbP2twzAQANZJjpX9g8usqySbO5PTsbPZuYVDK+vbu2Rt5ZXE0AAxdDoc5vK4fO7UKYRdJRXxzBavS4sABGEAgeS+P1YwFQAFEAEr4ADsawAjKDkWiMRMoQ9PABdaGw+HPRZglhiKDIeQAZTMMEsADoACLEMgwYDERz+QKuBlNektb7DAD6f1yktEiph6Qu+2QYDy6HlOvSdN1cCZaT2v3FeHsEssADkAQUzT4IIEAkFQvLmT4xBL0ABZQJkFHAPK4IrlIjEMBkIoR40ezz68p0ZBEKhSogei0MuBkVBwEREMSGASSdQ9bQ6aIwiPAUKmcxWTS9ByOANVGp1FIsAAMAA8ABxkACcQ7JdBRa0q9joeNJq8n+AAzMSNxOACwAgHEkVkLZkse7jeVNZjjYbj1ZnNaojOlV2zXoR2Ciyu929qCcFJwkiIDHiVTpuk7bRylHSdJzHBcyXQDYhyHDZJxRChd13dBJzWOhiSHc86FvNZCLJCgyTWCghxROgtjINNYgOY4iFOUJdyIAN0g1QFgVCIdOKVXFCRJclQlIZB0C4nwLUia0YXk6gDAEIA)

```ts
import { useWriteContract } from 'wagmi'

const { writeContract } = useWriteContract()

const switchContractABI = [
  {
    inputs: [
      {
        components: [
          { internalType: 'address', name: 'nftContract', type: 'address' },
          { internalType: 'uint256', name: 'tokenId', type: 'uint256' },
          { internalType: 'uint256', name: 'donorFee', type: 'uint256' },
          { internalType: 'bool', name: 'isERC721', type: 'bool' },
        ],
        internalType: 'struct Switch.DonationInput[]',
        name: '_donations',
        type: 'tuple[]',
      },
    ],
    name: 'donateBatchNFT',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
] as const

writeContract({
  abi: switchContractABI,
  address: '0x8a901fb2dBfE727279C3738a4FF7Da61843d2853',
  functionName: 'donateBatchNFT',
  args: [
    [
      {
        nftContract: '0x9982d1e50059bc44e92f7084f532081c12c0bf6a',
        tokenId: 4n,
        donorFee: 0n,
        isERC721: true,
      },
    ],
  ],
})
```

**Reply by Neaxic** [reply]
Thanks that fixed the issue. I'm currently using TS, but I didn't const assert, - ill look into that.
Much appreciated.
